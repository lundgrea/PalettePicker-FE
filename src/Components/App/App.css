* {
  margin: 0;
  padding: 0;
  box-sizing: border-box;
}

html {
  height: 100%;
  overflow: hidden;
}

body {
  /* background: url("../../assets/bobRoss.jpg") no-repeat center center fixed;
  background-size: cover; */
}

.App {
  display: flex;
}

