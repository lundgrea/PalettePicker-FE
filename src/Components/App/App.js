import React, { Component } from "react";
import Sidebar from "../Sidebar/Sidebar";
import Main from "../Main/Main";
import { 
  fetchAllFolders, 
  fetchAllPalettes, 
  fetchAPalette, 
  fetchAFolder, 
  fetchAFoldersPalettes, 
  deleteFolder, 
  deletePalette, 
  postNewPalette, 
  postNewFolder, 
  patchAPalette, 
  patchAFolder 
} from '../../apiCalls/apiCalls'
import "./App.css";

class App extends Component {
  constructor() {
    super();
    this.state = {
      folders: [],
      folderID: "",
      palettes: [],
      paletteID: "",
      currentPalette: [],
      isLoading: true,
      networkMessage: '',
      error: ''
    };
  }

  generateRandomColors = () => {
    let unlockedSwatches = this.state.currentPalette.map(swatch => {
      if (!swatch.isLocked) {
        return {color: hexCodeGen(), isLocked: false}
      }
      return swatch
    })

    let currentPalette = Array(5).fill({}).map(el => {
      return {color: hexCodeGen(), isLocked: false}
    })

    function hexCodeGen () {
      return "#" + Math.floor(Math.random() * 16777215).toString(16);
    }

    return this.state.currentPalette.length === 5 ? this.setState({currentPalette: unlockedSwatches}) : this.setState({currentPalette})

  }

  toggleLock = (e, color) => {
    e.preventDefault();
    const newArray = this.state.currentPalette.map(palette => {
      if (palette.color === color) {
        palette.isLocked = !palette.isLocked;
        return palette;
      } else {
        return palette;
      }
    });
    this.setState({ currentPalette: newArray });
  };

  clearFolderID = (e) => {
    e.preventDefault()
    this.setState({folderID: ""})
  }

  componentDidMount = () => {
    this.generateRandomColors();
    this.getAllFolders();
    // this.deleteAPalette(30)
    // this.deleteAFolder(9);
    // this.grabAPalette(16)
    // this.addNewPalette('Jolly', array[0], array[1], array[2], array[3], array[4], 11)
    // this.addNewFolder('Jambo Rambo')
    // this.updateAPalette(19, 'name', 'Bizarre Violet')
    // this.updateAFolder(11, 'Doozy Boozy')
  };

  getAFolder = folderId => {
    fetchAFolder(folderId)
    .then(folder => this.setState({ folderID: folder[0].id }))
    .catch(error => this.setState({error}))
  }

  getAllFolders = () => {
    fetchAllFolders()
    .then(folders => this.setState({ folders }))
    .then(isLoading => this.setState({ isLoading: false }))
    .catch(error => this.setState({error: "Happy Little Accident While Getting All Folders! Please Try Again."}))
  }

  getAPalette = paletteId => {
    fetchAPalette(paletteId)
    .then(currentPalette => this.setState({ currentPalette }))
    .catch(error => this.setState({error}))
  }

  getAFoldersPalettes = (e, folderID) => {
    e.preventDefault()
    fetchAFoldersPalettes(folderID)
    .then(palettes => this.setState({ folderID, palettes }))
    .catch(error => this.setState({error}))
  }

  getAllPalettes = () => {
    fetchAllPalettes()
    .then(palettes => this.setState({ palettes }))
    .catch(error => this.setState({error}))
  }

  addNewPalette = (name, c1, c2, c3, c4, c5, folderId) => {
    postNewPalette(name, c1, c2, c3, c4, c5, folderId)
    .then(networkMessage => this.setState({networkMessage}))
    .then(() => this.getAllFolders())
    .catch(error => this.setState({error: "Happy Little Accident Adding Your Palette!"}))
  }

  addNewFolder = folderName => {
    postNewFolder(folderName)
    .then(networkMessage => this.setState({networkMessage}))
    .then(() => this.getAllFolders())
    .catch(error => this.setState({error: "Happy Little Accident Adding Your Folder! "}))
  }

  updateAPalette = (paletteId, value) => {
    patchAPalette(paletteId, value)
   .then(networkMessage => this.setState({networkMessage}))
    .catch(error => this.setState({error}))  
  }

  updateAFolder = (folderId, newName) => {
    patchAFolder(folderId, newName)
    .then(networkMessage => this.setState({networkMessage}))
    .catch(error => this.setState({error}))
  }

  deleteAPalette = paletteId => {
    deletePalette(paletteId)
    .then(networkMessage => this.setState({networkMessage}))
    .catch(error => this.setState({error}))
  }

  deleteAFolder = folderId => {
    deleteFolder(folderId)
    .then(networkMessage => this.setState({networkMessage}))
    .catch(error => this.setState({error}))
  }
 

  render() {
    return (
      <main className="App">
        {!this.state.isLoading 
        && 
        <Sidebar 
        folders={this.state.folders} 
        addNewFolder={this.addNewFolder}
        getAllFolders={this.getAllFolders}
        getAFoldersPalettes={this.getAFoldersPalettes}/>}
        <Main
          currentPalette={this.state.currentPalette}
          generateRandomColors={this.generateRandomColors}
          toggleLock={this.toggleLock}
          palettes={this.state.palettes}
          folderID={this.state.folderID}
          clearFolderID={this.clearFolderID}
          folders={this.state.folders}
          isLoading={this.state.isLoading}
          addNewPalette={this.addNewPalette}
        />
      </main>
    );
  }
}

export default App;
