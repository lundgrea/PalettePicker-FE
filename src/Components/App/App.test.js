import React from 'react';
import { shallow } from "enzyme";
import { App }  from './App';
import { configure } from "enzyme";
import Adapter from "enzyme-adapter-react-16";

configure({ adapter: new Adapter() });


let wrapper;

beforeEach(() => {
  wrapper = shallow(<App />)

  // fetchAllFolders.mockImplementation(() => {
  //   return Promise.resolve({folder: [{id: 1, name: 'ColorRama'}, {id: 2, name: 'Floss'}]}) 
  // })
});

describe("App", () => {
  it("should match snapshot", () => {
    expect(wrapper).toMatchSnapshot();
  });


  it('should start with empty state, besides those updated onLoad', () => {
    expect(wrapper.state('folderID')).toEqual("");
    expect(wrapper.state('palettes')).toEqual([]);
    expect(wrapper.state('paletteID')).toEqual("");
    expect(wrapper.state('isLoading')).toEqual(true);
    expect(wrapper.state('networkMessage')).toEqual("");
    expect(wrapper.state('error')).toEqual("");
  })

  it('should generateRandomColors on pageLoad is called', () => {
    expect(wrapper.state('currentPalette').length).toEqual(5);
  })

  it("should reset state to an empty string", () => {
    wrapper.instance().setState({folderID: 878})
    const mockClickEvent = { preventDefault: jest.fn() };
    wrapper.instance().clearFolderID(mockClickEvent)
    expect(wrapper.state("folderID")).toEqual("");
  });

  it('should call generateRandomColors and getAllFolder when componentDidMount is called', () => {
    wrapper.instance().getAllFolders = jest.fn();
    wrapper.instance().generateRandomColors = jest.fn()
    wrapper.instance().componentDidMount();
    expect(wrapper.instance().getAllFolders).toHaveBeenCalled();
    expect(wrapper.instance().generateRandomColors).toHaveBeenCalled();
  });

})


