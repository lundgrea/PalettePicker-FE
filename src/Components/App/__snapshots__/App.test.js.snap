// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should match snapshot 1`] = `
<main
  className="App"
>
  <Main
    addNewPalette={[Function]}
    clearFolderID={[Function]}
    currentPalette={
      Array [
        Object {
          "color": "#e310e9",
          "isLocked": false,
        },
        Object {
          "color": "#0c6f54",
          "isLocked": false,
        },
        Object {
          "color": "#c2191f",
          "isLocked": false,
        },
        Object {
          "color": "#98c5aa",
          "isLocked": false,
        },
        Object {
          "color": "#987ec5",
          "isLocked": false,
        },
      ]
    }
    deleteAPalette={[Function]}
    folderID=""
    folders={Array []}
    generateRandomColors={[Function]}
    isLoading={true}
    palettes={Array []}
    toggleLock={[Function]}
  />
</main>
`;
